client-misk-dynamodb-v2/src/main/kotlin/app/cash/backfila/client/misk/dynamodbv2/internal/keyranges.kt

```diff
@@ -0,0 +1,191 @@
+    )
+  }
+
+  internal fun decodeSegment(segment: ByteString): SegmentData {
+    val buffer = Buffer().write(segment)
+    val tag = buffer.readUtf8(2)
+    require(tag == VERSION) {
+      "Encountered an incorrect version $tag instead of $VERSION . Make sure any deploys have " +
+          "completed and re-create the backfill."
+    }
+    require(buffer.readUtf8(1) == ":")
+    val offset = buffer.readUtf8(10).toInt()
+    require(buffer.readUtf8(1) == "/")
+    val count = buffer.readUtf8(10).toInt()
+    var lastEvaluatedKey: Map<String, AttributeValue>? = null
+    if (!buffer.exhausted()) { // has a lastEvaluatedKey
+      require(buffer.readUtf8(1) == ":")
+      lastEvaluatedKey = adapter.fromJson(buffer)
+    }
+    require(buffer.exhausted())
+    return SegmentData(offset, count, lastEvaluatedKey)
+  }
+
+  internal fun encodeKeyRange(
+    start: Int,
+    end: Int,
+    count: Int,
+    lastEvaluatedKey: Map<String, AttributeValue>? = null
+  ): KeyRange {
+    return KeyRange.Builder()
+      .start(encodeSegment(start, count, lastEvaluatedKey))
+      .end(encodeSegment(end, count))
+      .build()
+  }
+
+  private fun encodeSegment(
+    offset: Int,
+    count: Int,
+    lastEvaluatedKey: Map<String, AttributeValue>? = null
+  ): ByteString {
+    require(offset in 0..count)
+    val stringOffset = INT_FORMAT.format(offset)
+    val stringCount = INT_FORMAT.format(count)
+    val buffer = Buffer()
+      .writeUtf8("$VERSION:$stringOffset/$stringCount")
+    if (lastEvaluatedKey != null) {
+      buffer.writeUtf8(":").writeUtf8(adapter.toJson(lastEvaluatedKey))
+    }
+    return buffer.readByteString()
+  }
+
+  internal data class SegmentData(
+    val offset: Int,
+    val count: Int,
+    val lastEvaluatedKey: Map<String, AttributeValue>?
+  )
+
+  internal data class AttributeValueJson(
+    var s: String? = null,
+    val n: String? = null,
+    val b: ByteString? = null,
+    val sS: List<String>? = null,
+    val nS: List<String>? = null,
+    val bS: List<ByteString>? = null,
+    val m: Map<String, AttributeValueJson>? = null,
+    val l: List<AttributeValueJson>? = null,
+    val nULLValue: Boolean? = null,
+    val bOOL: Boolean? = null
+  )
+}
+
+object AwsAttributeValueAdapter {
+  @ToJson
+  internal fun toJson(attributeValue: AttributeValue): DynamoDbKeyRangeCodec.AttributeValueJson {
+    return DynamoDbKeyRangeCodec.AttributeValueJson(
+      attributeValue.s(),
+      attributeValue.n(),
+      attributeValue.b()?.asByteArray()?.toByteString(),
+      attributeValue.ss(),
+      attributeValue.ns(),
+      attributeValue.bs()?.map { it.asByteArray().toByteString() },
+      attributeValue.m()?.mapValues { toJson(it.value) },
+      attributeValue.l()?.map { toJson(it) },
+      attributeValue.nul(),
+      attributeValue.bool()
+    )
+  }
+
+  @FromJson
+  internal fun fromJson(json: DynamoDbKeyRangeCodec.AttributeValueJson): AttributeValue {
+    return AttributeValue.builder()
+      .let {
```

Comment:
In dynamo v1 you could just set all of these and it would work. In v2 it's an error to set more than one value. I'm relying on null checks and null+empty checks to do this properly.